pagefaultgames/pokerogue
# [Testing] Flexible Testing Wrapper for Phaser-Based Battle-Scenes

## What are the changes?
This PR introduces a wrapper for Phaser.Game to allow testing any project feature without worrying about the "drawing" or "rendering" parts. While testing some graphical aspects of the game would be a plus, it's not the main focus at the moment. With this wrapper, we can now instantiate a Battle-Scene and test phase by phase what we want. 

## Why am I doing these changes?
These changes are introduced to facilitate testing by allowing manipulation of data between phases to test different scenarios. It also enables intercepting displayed messages, pausing the game between phases, executing code during a prompt, and sending Button-type inputs as if in the game. This improvement stems from the need to enhance the testing capabilities for the project, providing a more controlled and flexible testing environment.

## What did change?
- The introduction of a Phaser.Game wrapper allows for:
  - Instantiation of a Battle-Scene for phase-by-phase testing.
  - Data manipulation between phases.
  - Message interception.
  - Game pausing between phases.
  - Execution of code during prompts.
  - Simulation of Button-type inputs.
- Mocking fetches to retrieve assets from the public folder.
- Interception of console.log outputs.
- Mocking of all drawing classes (text, rectangle, sprite, container, polygon, image, nineslice, graphics) to ensure the proper functioning of certain game actions.
- A wrapper for local storage to simulate it in tests.
- A wrapper for Phaser.Time.clock to update the clock every 100ms instead of during pre-update and update of the scene, as the Phaser engine is simulated and doesn't actually exist.

## Why are not all tests passing?
Because they are detecting bugs that need to be resolved, but it is not the purpose of this PR to fix them.

### Screenshots/Videos
![image](https://github.com/pagefaultgames/pokerogue/assets/44787002/b90ccce3-ed30-41c0-967f-8d502bff71c9)


## How to test the changes?
To test these changes, reviewers can check out the branch and run the provided rudimentary tests which include:
- Ability tests (intimidate and moxie)
- HeldItems test (toxic orb)
- Achievement tests
- Input tests with a fake gamepad/keyboard/mobile
- Move tests (Tackle, Growth, Tail Whip)
- Starter selection screen tests
- Specific battle tests like execution order when multiple Pokémon have the same speed

Ensure that when testing or stopping at phases not yet intercepted, they are added to PhaseInterceptor.

## Checklist
